controller/src/controller/utils/stimulation.py

```diff
@@ -107,3 +140,84 @@ def chunk_protocols_in_stim_info(
         max_subprotocol_idx_counts[protocol_id] = tuple(original_idx_counts)
 
     return chunked_stim_info, subprotocol_idx_mappings, max_subprotocol_idx_counts
+
+
+def _check_subprotocol_type(subprotocol: dict[str, Any], protocol_id: int, idx: int) -> Any:
+    subprotocol["type"] = subprotocol_type = subprotocol["type"].lower()
+    # validate subprotocol type
+    if subprotocol_type not in VALID_SUBPROTOCOL_TYPES:
+        raise WebsocketCommandError(
+            f"Protocol {protocol_id}, Subprotocol {idx}, Invalid subprotocol type: {subprotocol_type}"
+        )
+
+    return subprotocol_type
+
+
+def validate_stim_subprotocol(
+    subprotocol: dict[str, Any], stim_type: str, protocol_id: int, idx: int
+) -> None:
+    subprotocol_type = _check_subprotocol_type(subprotocol, protocol_id, idx)
+
+    if subprotocol_type == "loop":
+        for nested_subprotocol in subprotocol["subprotocols"]:
+            validate_stim_subprotocol(nested_subprotocol, stim_type, protocol_id, idx)
+    else:
+        # validate subprotocol components
+        if subprotocol_type == "delay":
+            # make sure this value is not a float
+            subprotocol["duration"] = int(subprotocol["duration"])
+            total_subprotocol_duration_us = subprotocol["duration"]
+        else:  # monophasic and biphasic
+            max_abs_charge = (
+                STIM_MAX_ABSOLUTE_VOLTAGE_MILLIVOLTS
+                if stim_type == "V"
+                else STIM_MAX_ABSOLUTE_CURRENT_MICROAMPS
+            )
+
+            subprotocol_component_validators = {
+                "phase_one_duration": lambda n: n > 0,
+                "phase_one_charge": lambda n: abs(n) <= max_abs_charge,
+                "postphase_interval": lambda n: n >= 0,
+            }
+            if subprotocol_type == "biphasic":
+                subprotocol_component_validators.update(
+                    {
+                        "phase_two_duration": lambda n: n > 0,
+                        "phase_two_charge": lambda n: abs(n) <= max_abs_charge,
+                        "interphase_interval": lambda n: n >= 0,
```

Comment:
since these lambdas are all the same as the previous validator you could pull them out as variables instead of duplicating them

```python
phase_duration_validator = lambda n: n > 0
phase_charge_validator = lambda n: abs(n) <= max_abs_charge
interphase_validator = lambda n: n >= 0

subprotocol_component_validators = {
    "phase_one_duration": phase_duration_validator,
    ...
}

...

subprotocol_component_validators.update({
    "phase_two_duration": phase_duration_validators,
   ...
})
```python

Reply:
one of the pre-commit hooks complains about this for some reason